compiler/AST/iterator.cpp

```diff
@@ -455,11 +471,24 @@ buildZip1(IteratorInfo* ii, Vec<BaseAST*>& asts, BlockStmt* singleLoop) {
     zip1body->insertAtTail(new DefExpr(condTemp));
     zip1body->insertAtTail(new CallExpr(PRIM_MOVE, condTemp,
                                         new CallExpr(PRIM_CAST, moreType->symbol, condExpr)));
-    zip1body->insertAtTail(new CallExpr(PRIM_SET_MEMBER, ii->zip1->_this,
+    zip1body->insertAtTail(new CallExpr(PRIM_SET_MEMBER, my_this,
                                         ii->iclass->getField("more"), condTemp));
-  } else if (isCForLoop(singleLoop)) {
+  } else if (singleLoop->isCForLoop()) {
     // CForLoops do not use the "more" field in their loop termination test.
     // See the code for that special case in buildHasMore().
+  } else if (singleLoop->isForLoop()) {
+    // Get the underlying iterator.
+    INT_ASSERT(my_this->type == ii->iclass);
+    Symbol* iteratorField = ii->iclass->getField("F1__iterator");
+    VarSymbol* iterator = newTemp("forloop_iter", iteratorField->type);
+    zip1body->insertAtTail(new DefExpr(iterator));
+    CallExpr* iteratorFieldRead = new CallExpr(PRIM_GET_MEMBER_VALUE, my_this,
+                                               iteratorField);
+    zip1body->insertAtTail(new CallExpr(PRIM_MOVE, iterator, iteratorFieldRead));
+
+    // Call the contained iteratator's zip1() function. 
+    IteratorInfo* under_ii = getIteratorInfoForIterator(iterator);
+    zip1body->insertAtTail(new CallExpr(under_ii->zip1, iterator));
   } else {
```

Comment:
Should remove the "ForLoops should have been replaceed in expandForLoop()." from here and buildzip3